sklearn/cluster/tests/test_birch.py

```diff
@@ -228,3 +229,18 @@ def test_feature_names_out():
 
     names_out = brc.get_feature_names_out()
     assert_array_equal([f"birch{i}" for i in range(n_clusters)], names_out)
+
+
+def test_transform_match_across_dtypes():
+    X, _ = make_blobs(n_samples=80, n_features=4, random_state=0)
+    brc = Birch(n_clusters=4)
+    Y_64 = brc.fit_transform(X)
+    Y_32 = brc.fit_transform(X.astype(np.float32))
```

Comment:
```suggestion
    Y_32 = brc.fit_transform(X.astype(np.float32, copy=False))
```

Reply:
Since `X` starts off as a `float64`, I think using `astype` will always copy even if `copy` is False?

```python
import numpy as np

X = np.asarray([1.1, 2.3], dtype=np.float64)
X_32 = X.astype(np.float32, copy=False)
X_32[0] = 6.1

print(X[0])
# 1.1
```

Reply:
ah right. It's a reflex each time I see `astype` :smile: